### ISSUE TITLE
[Bug]  AITER `gemm_a4w4` MXFP4 f4gemm returns garbage when contraction dim K is not a multiple of 256

### ISSUE THREAD

**Issue by mvstrauss** [ROCm/aiter#4176]
## Summary

The AITER MXFP4 A4W4 GEMM (`aiter.gemm_a4w4`, gfx950 asm / CK f4gemm path)
produces **silently wrong results** (relative error ~10^3–10^7) whenever the
**contraction dimension K is not a multiple of 256**. There is no error, no
warning, and no fallback — the kernel just returns garbage. When `K % 256 == 0`
the result is correct (MXFP4 quantization noise only, rel-err ~0.16 on random
operands).

The wrapper pads and asserts only the **output row** dimension to a multiple of
32; it applies **no alignment guard or tail handling to K**. MXFP4 quantization
pads K only to a multiple of the block size (32), so any K that is `%32` but not
`%256` (e.g. 1664, 1824, 1856) reaches the kernel and is mishandled.

This is content-independent (reproduces on random Gaussian operands) and
backend-specific: hipBLASLt's FP4 GEMM computes the same inputs correctly
(it requires `K % 128`), which is how we isolated the fault to the AITER kernel.

## Environment

- GPU: AMD Instinct MI355X (gfx950)
- ROCm: 7.x
- AITER: `module_gemm_a4w4_asm` / `module_gemm_a4w4_blockscale` (a4w4 f4gemm)
- primus_turbo: 0.2.0 (used only for the MXFP4 quantizer + scale/weight shuffle)
- PyTorch: `torch.float4_e2m1fn_x2` (E2M1, block-32, E8M0 scale)
- Recipe: MXFP4 MX_BLOCKWISE, block_size=32, NT layout, `bpreshuffle=True`

## Where it happens

`aiter/ops/gemm_op_a4w4.py::gemm_a4w4` (gfx950 branch):

```python
m = A.numel() // A.shape[-1]
n = B.shape[0]
k = A.shape[-1] * 2
...
out = torch.empty(((m + 31) // 32 * 32, n), dtype=dtype, device=A.device)  # output row pad -> %32
ck_config = get_GEMM_config(m, n, k)      # kernel selected by (m, n, k)
...
assert out.shape[0] % 32 == 0, "Dim0 of gemm_a4w4_asm output needs to be padded to multiples of 32!"
gemm_a4w4_asm(A.view(m, k // 2), B, A_scale, B_scale, out, ...)   # no K alignment guard
```

- The **output row** dim `m` is padded to `%32` and asserted.
- **K is never checked.** The selected asm/CK f4gemm kernel appears to assume K
  is a multiple of the MFMA K-tile × pipeline depth (256 for MXFP4/E2M1 packing);
  a K tail of 32/64/128/… is read past/misaligned, corrupting the whole output.
- `get_GEMM_config(m, n, k)` may also return a config tuned for a different K,
  compounding the issue.

## Minimal reproducer

See `aiter_f4gemm_ktail_repro.py` (accompanying this report). It quantizes random
bf16 operands to MXFP4 with primus_turbo's standard quantizer and calls
`aiter.gemm_a4w4` exactly as primus_turbo's `GEMMFP4AITERBackend` does
(`shuffle_scale` + `shuffle_weight` + `bpreshuffle=True`), sweeping K with fixed
output dims M=N=512 (both `%256`).

Observed output (MI355X / gfx950):

```
AITER a4w4 MXFP4 f4gemm  |  M=512 N=512, sweeping contraction K
      K  K%256       rel_err  status
   1536      0     1.575e-01  OK
   1664    128     3.473e+03  >>> GARBAGE
   1792      0     1.579e-01  OK
   1824     32     3.337e+03  >>> GARBAGE
   1856     64     3.307e+03  >>> GARBAGE
   2048      0     1.575e-01  OK
   2304      0     1.580e-01  OK
   3072      0     1.584e-01  OK
```

`rel_err ~0.16` is the expected MXFP4 quant error; `rel_err ~10^3` is garbage.
The pattern is exactly `K % 256`: correct iff `K % 256 == 0`, regardless of
operand content (random data), regardless of M/N (both fixed and `%256`).

## Real-world impact

This kernel bug causes a **training-time gradient explosion** for MXFP4 diffusion
training (Wan2.2-T2V-A14B on MI355X). Of a Linear's three GEMMs, only the weight
gradient (Wgrad, `grad_W = grad_out^T @ x`) contracts over the **token**
dimension; forward and dgrad contract over feature dims (in/out_features), which
are `%256` for this model and therefore correct. The token count is `seq*batch`;
for cross-attention **key/value** projections it is `text_seq * batch`
(e.g. 226 * 8 = 1808, `%256 = 16`). Those Wgrad GEMMs hit the K-tail bug and
return ~10^7× too-large weight gradients, blowing up the global grad norm to ~10^6
at step 0 and preventing convergence. Layers whose token count happens to be
`%256` (self-attention over 512p video latents, or FLUX at seq 256) are unaffected
— which is why the failure looked data/model-dependent until isolated to K%256.

## Suggested fixes (in the kernel/wrapper)

1. **Add a K-alignment guard** in `gemm_a4w4` (fail loudly instead of silently
   returning garbage), e.g. `assert k % 256 == 0` for the MXFP4 asm/CK path, or
2. **Handle the K tail** in the asm/CK f4gemm kernels (mask/predicate the final
   partial K-tile), or
3. **Pad K internally** in the wrapper: zero-pad the packed operands and their
   E8M0 scales up to `ceil(K/256)*256` before dispatch and run the padded GEMM
   (zeros are exact for a contraction), mirroring the existing output-row padding.
4. At minimum, **document** the `K % 256` requirement alongside the existing
   `M % 16` / `N % 16` constraints so callers can pad.

## Current workaround (framework side)

Zero-pad the token/contraction dimension to a multiple of 256 before the Wgrad
GEMM (exact, since padded rows contribute 0 to `grad_out^T @ x`). This keeps the
fast preshuffled AITER path for all three GEMMs and restores correct gradients
(Wgrad amplification 6.4e7× → ~1.0×). Routing only Wgrad to hipBLASLt also works
numerically but its rocRoller FP4 solutions require the token dim `%256` on
fwd/dgrad and hang, so padding-for-AITER is preferred.

[aiter_f4gemm_ktail_repro.py](https://github.com/user-attachments/files/29876562/aiter_f4gemm_ktail_repro.py)

**Comment by valarLip** [comment]
i don't know where this shuffle comes from in your repro
shuffle_scale = torch.ops.primus_turbo_cpp_extension.shuffle_scale
shuffle_weight = torch.ops.primus_turbo_cpp_extension.shuffle_weight

**Comment by mvstrauss** [comment]
@valarLip Good catch, sorry for that extra dependency! Those or Primus-Turbo C++ extension ops, registered on import primus_turbo (line 23 of the repro).

To remove any doubt about the shuffle, I've attached a self-contained repro that uses only AITER's own utilities — aiter.get_triton_quant(aiter.QuantType.per_1x32) and aiter.ops.shuffle.shuffle_weight(..., layout=(16,16)), following your op_tests/test_gemm_a4w4.py. It shows the same result: correct iff K % 256 == 0, garbage otherwise, on random operands, with M=N=512 fixed. Note the preshuffle is identical across the sweep (only K varies), and hipBLASLt computes the same inputs correctly (K%128), so the fault is isolated to the a4w4 f4gemm K-tail handling, not the shuffle.

[aiter_f4gemm_ktail_repro_aiter_native.py](https://github.com/user-attachments/files/29893484/aiter_f4gemm_ktail_repro_aiter_native.py)

**Comment by valarLip** [comment]
We ran your attached AITER-native repro (`aiter_f4gemm_ktail_repro_aiter_native.py`) **verbatim** on MI355X (gfx950), AITER commit `a83fc86e6`. We **cannot reproduce the silent K%256 garbage**.

**Output of your exact native repro (M=N=512):**
```
      K  K%256       rel_err  status
   1536      0     1.575e-01  OK
   1664    128     1.580e-01  OK          <-- your report: 3.473e+03 GARBAGE
   1792      0     1.579e-01  OK
   1824     32           n/a  ERROR: 912 % 32 == 16
   1856     64     1.581e-01  OK          <-- your report: 3.307e+03 GARBAGE
   2048      0     1.575e-01  OK
   2304      0     1.580e-01  OK
   3072      0     1.584e-01  OK
```

**On the `K=1824` row — that `ERROR:` is not a kernel error, it is an `assert`.** The `ERROR:` prefix is just the repro's own `except Exception` print label; the actual exception is a deliberate **`AssertionError`** raised by `shuffle_weight(layout=(16,16))` at `aiter/ops/shuffle.py:165`:

```python
IN, IK = layout          # (16, 16)
BK = IK * 2              # = 32
assert x.shape[-1] % BK == 0, f"{x.shape[-1]} % {BK} == {x.shape[-1] % BK }"
```

`x` is the packed weight `b_fp4` of shape `[N, K/2] = [512, 912]` (2 fp4 elems/byte, `1824/2 = 912`), and `912 % 32 = 16 != 0`, so this shape guard fires **before `gemm_a4w4`/the kernel is ever called**. So `K=1824` is intentionally rejected by an `assert`, not silently miscomputed.

**Comment by mvstrauss** [comment]
@valarLip  Thank you very much for running this! This was very helpful, as unfortunately I didn't directly run the AITER native script myself but only translated the one I did run myself, the one using the Primus-Turbo shuffle ops. This bug report is a misattribution and can now be closed!

The garbage we saw comes from our production stack (Primus-Turbo). It feeds the kernel via its own preshuffle ops (primus_turbo_cpp_extension.shuffle_scale/shuffle_weight that you asked about) plus its own MXFP4 quantizer, and that path only gates K%32 (AITER_FP4GEMM_K_MULTIPLE = 32). The tell is K=1824: AITER's shuffle_weight asserts and rejects it, whereas Primus-Turbo's silently proceeds and returns garbage for the same shape — so the misaligned layout for K∤256 is on the Primus-Turbo preshuffle side, not your kernel. Padding K to %256 realigns it, which matches our fix. Apologies for the noise on the AITER tracker — this belongs in Primus-Turbo; please feel free to close. I'll refile there.

**Comment by carlushuang** [comment]
Not aiter issue
